packages/dashboard/src/contexts/DashContext/types/Schema.ts

```diff
@@ -0,0 +1,16 @@
+import type { DBSchema } from "idb/with-async-ittr";
+import type { Compilation } from "@truffle/compile-common";
+
+export interface Schema extends DBSchema {
+  Compilation: {
+    key: string;
+    value: {
+      dataHash: string;
+      data: Compilation;
+      timeAdded: number;
+    };
+    indexes: {
+      TimeAdded: number;
+    };
+  };
+}
```

Comment:
Have you considered how we'll update this schema in the future? For example, let's assume that we decide to release this with the compilation result matching and cache invalidation logic as-is, but then later on we want to come back in the future and improve it. How will we detect the version change? Will we migrate users forward, or will we just wipe their data and ask them to recompile the contract versions that they care about?

I think at a minimum you might want to consider adding a schema version field here so that we can easily detect what version users are on, report on the schema version via error messages, analytics, etc if necessary, etc. It'll also serve as a hint to future devs that they need to consider existing installs when updating the schema.

Reply:
That said, a lazy version detection and data migration strategy is totally valid in a lot of cases, so if you already considered that and want to opt to go that way, that's totally fine by me.

Reply:
Yes the schema is [versioned](https://github.com/trufflesuite/truffle/blob/35a9980fa901c2d1add466f0927541ae44d8abed/packages/dashboard/src/contexts/DashContext/state.ts#L18). And `idb` provides a nice way to run an `upgrade` function when the code specifies a version higher than that in the browser, to which you can add transactions to. I think ideally migrations in the future preserve as much compilation (and other) data as possible, and wipe only as the last resort if we cannot guarantee integrity.